File: algorithms/topological_sorting.java
// Topological Sorting is used to order the vertices in a directed graph such that for every directed edge `u -> v`, vertex `u` comex before vertex `v` in the ordering.
// This ordering is possible if the graph is a Directed Acyclic Graph (DAG).

// Consider the course prerequisites in a university.
// Each course has dependencies (prerequisites) that must be completed before taking the course.
// Topological sorting can be used to determine the order of courses to be taken, ensuring that all prerequisites are fulfilled.

import java.util.*;

class Graph {
    private int vertices;
    private List<List<Integer>> adjacencyList;

    public Graph(int vertices) {
        this.vertices = vertices;
        this.adjacencyList = new ArrayList<>();
        for (int i = 0; i < vertices; i++) {
            adjacencyList.add(new ArrayList<>());
        }
    }

    public void addEdge(int u, int v) {
        adjacencyList.get(u).add(v);
    }

    public List<Integer> topologicalSort() {
        Stack<Integer> stack = new Stack<>();
        boolean[] visited = new boolean[vertices];

        for (int i = 0; i < vertices; i++) {
            if (!visited[i]) {
                dfs(i, visited, stack);
            }
        }

        List<Integer> topologicalOrder = new ArrayList<>();
        while (!stack.isEmpty()) {
            topologicalOrder.add(stack.pop());
        }

        return topologicalOrder;
    }

    private void dfs(int vertex, boolean[] visited, Stack<Integer> stack) {
        visited[vertex] = true;
        for (int neighbor : adjacencyList.get(vertex)) {
            if (!visited[neighbor]) {
                dfs(neighbor, visited, stack);
            }
        }
        stack.push(vertex);
    }
}

File: algorithms/topological_sorting.kt
// Topological Sorting of a directed graph is a linear ordering of its vertices such that for every directed edge UV from vertex U to vertex V, U comes before V in the ordering.
// This algorithm is used in scenarios where there is a dependency between tasks or components.
// It's important to note that topoplogical sorting is possible only if the graph has no directed cycles, i.e., if it's a Directed Acyclic Graph (DAG).

// Imagine you're a project manager handling a complex project with many interdependent tasks.
// Some tasks cannot begin until others are completed.
// For instance, you can't start painting a house until the construction is finished.
// In such scenarios, determining an order in which undertake tasks while respecting their dependencies is crucial.
// This is where topoplogical sorting comes in handy.
